### PR TITLE
[DEV-2206] add webhook table

### PR DESCRIPTION
**Pull request by Yannick1712**
#### Pre-Release

- [ ] Check migrations
  - No database related infos (sqldb-xxx)
  - Impact on GS (new/removed columns)
- [ ] Check for linter errors (in PR)
- [ ] Test basic user operations (on [DFX services](https://dev.services.dfx.swiss))
  - Login/logout
  - Buy/sell payment request
  - KYC page

#### Post-Release

- Test basic user operations
- Monitor application insights log

### CODE REVIEW

**Comment by Yannick1712** [comment]
Add migration

File: src/subdomains/generic/user/services/webhook/webhook.entity.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+  @Column({ length: 256, nullable: true })
+  reason: string;
+
+  @Column({ type: 'datetime2', nullable: true })
+  sentDate: Date;
+
+  // References
+  @ManyToOne(() => UserData, { nullable: true })
+  user: User;
```
Comment (davidleomay) [inline src/subdomains/generic/user/services/webhook/webhook.entity.ts:23]: `UserData` copy&paste error?
-> redo migration

File: src/subdomains/generic/user/services/webhook/webhook-notification.service.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+  private async triggerUserDataWebhook<T extends PaymentWebhookData | KycWebhookData>(
+    userData: UserData,
+    data: T,
+    type: WebhookType,
+    reason?: string,
+  ): Promise<void> {
+    for (const user of userData.users) {
+      await this.triggerUserWebhook(user, data, type, reason);
+    }
+  }
```
Comment (davidleomay) [inline src/subdomains/generic/user/services/webhook/webhook-notification.service.ts:66]: I would create a webhook entity for every webhook to be sent (instead of looping here), so we can keep track of every single webhook. Then you could also remove the UserData field on the webhook.

File: src/subdomains/generic/kyc/services/kyc-admin.service.ts [hunk 3]
```diff
@@ -55,7 +55,7 @@ export class KycAdminService {
   async triggerWebhook(dto: KycWebhookTriggerDto): Promise<void> {
     const kycStep = await this.kycStepRepo.findOne({
       where: [{ id: dto.kycStepId }, { userData: { id: dto.userDataId } }],
-      relations: { userData: true },
+      relations: { userData: { users: { wallet: true } } },
       order: { updated: 'DESC' },
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/services/kyc-admin.service.ts:59]: I'm not sure if we should load all users with wallets everywhere. It will probably slow down everything. Maybe we can just load the users when we store the webhook?

File: src/subdomains/generic/user/services/webhook/webhook.service.ts [hunk 4]
```diff
@@ -1,103 +1,88 @@
...
+      where: {
+        data: dto.data,
+        type: dto.type,
+        reason: dto.reason,
+        user: { id: dto.user.id },
+        sentDate: IsNull(),
+      },
+      relations: { user: true },
     });
+    if (existing) throw new BadRequestException('Webhook already created');
```
Comment (davidleomay) [inline src/subdomains/generic/user/services/webhook/webhook.service.ts:82]: Do we need to check, if the wallet of the user has a webhook URL configured? Doesn't make sense to store the webhook otherwise ... WDYT?

File: src/subdomains/generic/user/services/webhook/webhook-notification.service.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+  // --- HELPER METHODS --- //
+
+  private async triggerUserWebhook<T extends PaymentWebhookData | KycWebhookData>(
+    user: User,
+    data: T,
+    type: WebhookType,
+    reason?: string,
+  ): Promise<void> {
```
Comment (davidleomay) [inline src/subdomains/generic/user/services/webhook/webhook-notification.service.ts:58]: You can just use the entity as a parameter here 😉

File: src/subdomains/generic/kyc/services/limit-request.service.ts [hunk 6]
```diff
@@ -69,7 +69,10 @@ export class LimitRequestService {
   }
 
   async updateLimitRequest(id: number, dto: UpdateLimitRequestDto): Promise<LimitRequest> {
-    const entity = await this.limitRequestRepo.findOne({ where: { id }, relations: ['userData'] });
+    const entity = await this.limitRequestRepo.findOne({
+      where: { id },
+      relations: { userData: { users: { wallet: true } } },
+    });
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/services/limit-request.service.ts:75]: Is this change required?

File: src/subdomains/generic/user/services/webhook/webhook.service.ts [hunk 7]
```diff
@@ -1,103 +1,103 @@
...
-  ) {}
+  constructor(private readonly webhookRepo: WebhookRepository, private readonly userRepo: UserRepository) {}
 
   async kycChanged(userData: UserData): Promise<void> {
-    await this.triggerUserDataWebhook(userData, WebhookDataMapper.mapKycData(userData), WebhookType.KYC_CHANGED);
+    if (!userData.users)
+      userData.users = await this.userRepo.find({
+        where: { userData: { id: userData.id } },
+        relations: { wallet: true },
+      });
```
Comment (davidleomay) [inline src/subdomains/generic/user/services/webhook/webhook.service.ts:22]: Could do `userData.users ??= ...`

File: src/subdomains/generic/user/services/webhook/webhook.service.ts [hunk 8]
```diff
@@ -1,103 +1,103 @@
...
+    }
   }
 
   async kycFailed(userData: UserData, reason: string): Promise<void> {
-    await this.triggerUserDataWebhook(userData, WebhookDataMapper.mapKycData(userData), WebhookType.KYC_FAILED, reason);
+    if (!userData.users)
+      userData.users = await this.userRepo.find({
+        where: { userData: { id: userData.id } },
+        relations: { wallet: true },
+      });
```
Comment (davidleomay) [inline src/subdomains/generic/user/services/webhook/webhook.service.ts:37]: Same here

File: src/subdomains/generic/user/services/webhook/webhook-notification.service.ts [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    private readonly http: HttpService,
+    private readonly notificationService: NotificationService,
+  ) {}
+
+  @Cron(CronExpression.EVERY_5_MINUTES)
+  @Lock(1800)
+  async checkCryptoPayIn() {
+    if (DisabledProcess(Process.WEBHOOK)) return;
+    await this.sendOpenWebhooks();
+  }
```
Comment (davidleomay) [inline src/subdomains/generic/user/services/webhook/webhook-notification.service.ts:31]: Haha, good method name 😆

File: src/subdomains/generic/user/services/webhook/webhook-notification.service.ts [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+      }
+    }
+  }
+
+  // --- HELPER METHODS --- //
+
+  private async triggerUserWebhook<T extends PaymentWebhookData | KycWebhookData>(webhook: Webhook): Promise<void> {
+    try {
+      if (!webhook.user.wallet.apiUrl) return;
+      if (!webhook.user.wallet.apiKey) throw new Error(`ApiKey for wallet ${webhook.user.wallet.name} not available`);
```
Comment (davidleomay) [inline src/subdomains/generic/user/services/webhook/webhook-notification.service.ts:56]: We should probably also throw an error, if apiUrl is not defined 🤔

Comment (Yannick1712) [reply src/subdomains/generic/user/services/webhook/webhook-notification.service.ts:56]: Yeah thought that too 👍

**Review by davidleomay — APPROVED** [review]

**Review by davidleomay — APPROVED** [review]
